Support bullet lists

anntzer opened the pull request:
The first commit reimplements the docstring handler to use a NodeVisitor.
The second commit adds support for bullet lists (#45).

Enumerated lists not implemented yet mostly because they're a pain as you need to handle all kinds of enumeration types (integer, alphabetic, roman) but can basically be implemented with a similar approach.

anntzer commented:
Actually I think a simpler solution may be to overwrite the rst directives to simply pass through the underlying source.

anntzer commented:
Turns out supporting enumerated lists was not that hard.

evanunderscore commented:
@anntzer I've added you as a collaborator to the project. I'm happy to review any code you'd like me to, but in general I trust your judgement.

anntzer commented:
Thanks!

Are you OK with the general approach of this PR, at least?

evanunderscore commented:
It looks like a good approach to me, but I don't have a whole lot of experience with docutils. For your test case, is it a deliberate choice to have the bullet points separated by two newlines? I would have expected one.

You might also be interested to look at [the approach clize took with this](https://github.com/epsy/clize/blob/b62697cd66465de8509a19bb42eb316ced60fccd/clize/help.py#L636). This is something it added relatively recently and I haven't had time to look at it in detail.

anntzer commented:
I had a quick look at clize and I don't think its doctree walker is anywhere close to supporting this, but I may wrong.

Adding a blank line between each entry is because at the doctree level, there is no way to know whether there was a blank line to start with or not, i.e. the following doctrees are similar except that for the second one, each list_item contains two paragraphs instead of one
```
- A
- simple
- list
```
```
- A long list

  Where each entry

- Has multiple

  Paragraphs.
```
Given that I think the second case looks more ugly if the middle blank line is removed than the first one does when extra blank lines are added, I decided to add blank lines everywhere.

(Of course we could add some heuristic looking at the number of paragraphs but that just seems wrong.)

Actually I lied a bit, paragraph nodes in the doctree have a `line` attribute indicating the line they start at so in theory it should be possible to add the number of physical (not logical) lines in the paragraph and figure out whether to insert newlines, but that'd be for another PR (plus you'd need to handle literal blocks etc.)

evanunderscore commented:
Fair enough. I'm happy for you to do this whichever way you see fit.

anntzer commented:
It was actually not too hard to keep track of how many newlines need to be inserted.  Fixed.

evanunderscore commented on test_defopt.py:
```diff
@@ -600,6 +600,39 @@ def func():
+
+1. bar
+
+2. baz""")
```
You can use `textwrap.dedent` to have this formatted sensibly in the code.

evanunderscore replied:
Or `inspect.cleandoc`.

anntzer commented:
A slightly ugly way to preserve multiple consecutive newlines (replace them by `"\n "` (note the space)) was proposed in http://bugs.python.org/issue31330#msg301164 and implemented here.
Switched to inspect.cleandoc (textwrap.dedent doesn't work, for a reason explained in a comment).

evanunderscore commented:
@anntzer Do you need anything more from me on this PR?

anntzer commented:
Sorry I sort of dropped the ball on this due to more pressing (real life) todos.  I think it's in a reasonable shape, so feel free to merge it if you're happy with it, or you can leave it as it is and I'll come back to it some day (doesn't look like there's much of a risk of merge conflicts given the amount of activity anyways :p).

evanunderscore commented:
No problem. Of course you're welcome to take as much time as you need, I just wanted to make sure you didn't need anything from me. I'll leave it to you to come back to whenever you have time.

anntzer commented:
I'm not a huge fan of how I basically hijacked your command line generator into a RST to ANSI-codes converters, and I think that if anything that RST->ANSI part could/should be factored into a separate library, but I guess that can wait and at least the PR does provide the originally intended behavior.

Let's merge as is and see if anything terrible happens :-)

evanunderscore commented:
The intent was always to do sensible things to render RST to the console, and I did have thoughts about making a separate library but decided to start simple. If things grow much more complex it might be worth reconsidering.

anntzer commented:
Any chance for a release some time soon?  Thanks.

evanunderscore commented:
Sorry, I completely forgot you had finished this. I'll try to sort out a release soon.

anntzer commented:
No worries at all!

evanunderscore commented:
Done, let me know if there are any problems.

anntzer commented:
thanks!
